### PR TITLE
fix(anvil): calculate trace address acording to the spec

### LINKED ISSUE #2663: `from` address incorrect for `delegatecall` in `trace_transaction`
### Component

Anvil

### Have you ensured that all of these are up to date?

- [X] Foundry
- [X] Foundryup

### What version of Foundry are you on?

forge 0.2.0 (63ed109 2022-08-09T00:20:50.211015Z)

### What command(s) is the bug in?

anvil --fork-block-number 15291050 -f https://eth-mainnet.alchemyapi.io/v2/$KEY

### Operating System

macOS (Apple Silicon)

### Describe the bug

I'm testing Anvil's `trace_transaction` by comparing it against the output of Nethermind using mainnet transactions.
The second call in tx [0x3255cce7312e9c4470e1a1883be13718e971f6faafb96199b8bd75e5b7c39e3a](https://etherscan.io/tx/0x3255cce7312e9c4470e1a1883be13718e971f6faafb96199b8bd75e5b7c39e3a) has a `delegatecall`.

From Nethermind, the `action.from` address is the contract that is sending the `delegatecall` and not the sender of the transaction in the previous call. That is, the `action.to` address of the first call should be the `action.from` address of the second call for this test tx.

You can also check out the Parity trace on Etherscan for this tx https://etherscan.io/vmtrace?txhash=0x3255cce7312e9c4470e1a1883be13718e971f6faafb96199b8bd75e5b7c39e3a&type=parity#raw

The first 2 calls from Anvil. The second call's `action.from` should be `0xe2f2a5c287993345a840db3b0845fbc70f5935a5` and not `0x2e4777139254ff76db957e284b186a4507ff8c67`
```json
[
        {
            "action": {
                "callType": "call",
                "from": "0x2e4777139254ff76db957e284b186a4507ff8c67",
                "gas": "0x1ea7b",
                "input": "0x43bcfab60000000000000000000000006b175474e89094c44da98b954eedeac495271d0f0000000000000000000000000000000000000000000000e0bd811c8769a824b00000000000000000000000000000000000000000000000e0ae9925047d8440b60000000000000000000000002e4777139254ff76db957e284b186a4507ff8c67",
                "to": "0xe2f2a5c287993345a840db3b0845fbc70f5935a5",
                "value": "0x0"
            },
            "blockHash": "0x57fc41478973665020c3da3e26dc0cad4e975bff441a24674e4f89c273046b54",
            "blockNumber": 15291051,
            "error": null,
            "result": {
                "gasUsed": "0x1ea7b",
                "output": "0x0000000000000000000000000000000000000000000000e0e82ca52ec6e6a4d3"
            },
            "subtraces": 1,
            "traceAddress": [
                0
            ],
            "transactionHash": "0xe3f418d4fb9ded6f13481b47ca19d3454b039c7ec96ba0778e920d39f9029a31",
            "transactionPosition": 0,
            "type": "call"
        },
        {
            "action": {
                "callType": "delegatecall",
                "from": "0x2e4777139254ff76db957e284b186a4507ff8c67",
                "gas": "0x1ce41",
                "input": "0x43bcfab60000000000000000000000006b175474e89094c44da98b954eedeac495271d0f0000000000000000000000000000000000000000000000e0bd811c8769a824b00000000000000000000000000000000000000000000000e0ae9925047d8440b60000000000000000000000002e4777139254ff76db957e284b186a4507ff8c67",
                "to": "0x15b2838cd28cc353afbe59385db3f366d8945aee",
                "value": "0x0"
            },
            "blockHash": "0x57fc41478973665020c3da3e26dc0cad4e975bff441a24674e4f89c273046b54",
            "blockNumber": 15291051,
            "error": null,
            "result": {
                "gasUsed": "0x1ce41",
                "output": "0x0000000000000000000000000000000000000000000000e0e82ca52ec6e6a4d3"
            },
            "subtraces": 2,
            "traceAddress": [
                0,
                1
            ],
            "transactionHash": "0xe3f418d4fb9ded6f13481b47ca19d3454b039c7ec96ba0778e920d39f9029a31",
            "transactionPosition": 0,
            "type": "call"
        }
```

The first two calls from Nethermind
```json
[
        {
            "action": {
                "callType": "call",
                "from": "0x2e4777139254ff76db957e284b186a4507ff8c67",
                "gas": "0x262b3",
                "input": "0x43bcfab60000000000000000000000006b175474e89094c44da98b954eedeac495271d0f0000000000000000000000000000000000000000000000e0bd811c8769a824b00000000000000000000000000000000000000000000000e0ae9925047d8440b60000000000000000000000002e4777139254ff76db957e284b186a4507ff8c67",
                "to": "0xe2f2a5c287993345a840db3b0845fbc70f5935a5",
                "value": "0x0"
            },
            "blockHash": "0xa47c8f1d8c284cb614e9c8e10d260b33eae16b1957a83141191bc335838d7e29",
            "blockNumber": 15291051,
            "result": {
                "gasUsed": "0x2131b",
                "output": "0x0000000000000000000000000000000000000000000000e0e82ca52ec6e6a4d3"
            },
            "subtraces": 1,
            "traceAddress": [],
            "transactionHash": "0x3255cce7312e9c4470e1a1883be13718e971f6faafb96199b8bd75e5b7c39e3a",
            "transactionPosition": 19,
            "type": "call"
        },
        {
            "action": {
                "callType": "delegatecall",
                "from": "0xe2f2a5c287993345a840db3b0845fbc70f5935a5",
                "gas": "0x23d88",
                "input": "0x43bcfab60000000000000000000000006b175474e89094c44da98b954eedeac495271d0f0000000000000000000000000000000000000000000000e0bd811c8769a824b00000000000000000000000000000000000000000000000e0ae9925047d8440b60000000000000000000000002e4777139254ff76db957e284b186a4507ff8c67",
                "to": "0x15b2838cd28cc353afbe59385db3f366d8945aee",
                "value": "0x0"
            },
            "blockHash": "0xa47c8f1d8c284cb614e9c8e10d260b33eae16b1957a83141191bc335838d7e29",
            "blockNumber": 15291051,
            "result": {
                "gasUsed": "0x1f6e1",
                "output": "0x0000000000000000000000000000000000000000000000e0e82ca52ec6e6a4d3"
            },
            "subtraces": 2,
            "traceAddress": [
                0
            ],
            "transactionHash": "0x3255cce7312e9c4470e1a1883be13718e971f6faafb96199b8bd75e5b7c39e3a",
            "transactionPosition": 19,
            "type": "call"
        }
```

### PR DESCRIPTION
**Pull request by mattsse**
## Motivation
Closes https://github.com/foundry-rs/foundry/issues/2656
Closes #2663

Fixes `traceAddress` calculation. this will add the index of the call in the parent node instead of the parent's depth

@naddison36 mind giving this another try, the from issue should be fixed as well

### CODE REVIEW

**Comment by naddison36** [comment]
Thanks for the quick turnaround @mattsse 

I'm new to Rust so I may have compiled and run your branch incorrectly, but when I retested I got this error

```
trace_transaction
thread 'tokio-runtime-worker' panicked at 'child exists in parent', anvil/core/src/eth/transaction/mod.rs:1076:18
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
[1]    83664 abort      /Users/nicholasaddison/.cargo/bin/anvil --fork-block-number 15291050 -f
```

After setting `RUST_BACKTRACE=full`, I get
```
trace_transaction
thread 'tokio-runtime-worker' panicked at 'child exists in parent', anvil/core/src/eth/transaction/mod.rs:1076:18
stack backtrace:
   0:        0x103345625 - __mh_execute_header
   1:        0x1030d805b - __mh_execute_header
   2:        0x103322bf8 - __mh_execute_header
   3:        0x10334894d - __mh_execute_header
   4:        0x1033496f0 - __mh_execute_header
   5:        0x103349274 - __mh_execute_header
   6:        0x1033491e9 - __mh_execute_header
   7:        0x1033491a5 - __mh_execute_header
   8:        0x1033c4d93 - __mh_execute_header
   9:        0x1030db17b - __mh_execute_header
  10:        0x1030db12c - __mh_execute_header
  11:        0x1033c4f09 - __mh_execute_header
  12:        0x102fb5048 - __mh_execute_header
  13:        0x102fb3dad - __mh_execute_header
  14:        0x102e560ca - __mh_execute_header
  15:        0x102e3e687 - __mh_execute_header
  16:        0x10303e5b7 - __mh_execute_header
  17:        0x10302ff4a - __mh_execute_header
  18:        0x10302d1ef - __mh_execute_header
  19:        0x10306d8ae - __mh_execute_header
  20:        0x103069df4 - __mh_execute_header
  21:        0x102dd5067 - __mh_execute_header
  22:        0x102dd4d97 - __mh_execute_header
  23:        0x102dd4c27 - __mh_execute_header
  24:        0x102dd876b - __mh_execute_header
  25:        0x102dd5067 - __mh_execute_header
  26:        0x102dd4d97 - __mh_execute_header
  27:        0x102dd4c27 - __mh_execute_header
  28:        0x102dd876b - __mh_execute_header
  29:        0x102dd5067 - __mh_execute_header
  30:        0x102dd4d97 - __mh_execute_header
  31:        0x102dda574 - __mh_execute_header
  32:        0x102df77a5 - __mh_execute_header
  33:        0x102dd5067 - __mh_execute_header
  34:        0x102dd4d97 - __mh_execute_header
  35:        0x102df8ee1 - __mh_execute_header
  36:        0x102df9efa - __mh_execute_header
  37:        0x102dd5067 - __mh_execute_header
  38:        0x102db9969 - __mh_execute_header
  39:        0x102d8115f - __mh_execute_header
  40:        0x102db3a82 - __mh_execute_header
  41:        0x102db25a2 - __mh_execute_header
  42:        0x103360978 - __mh_execute_header
  43:        0x10335f9b4 - __mh_execute_header
  44:        0x103128dd8 - __mh_execute_header
  45:        0x103364727 - __mh_execute_header
  46:        0x1033658f4 - __mh_execute_header
  47:        0x10334a389 - __mh_execute_header
  48:     0x7ff81a22d4e1 - __pthread_start
[1]    83811 abort      /Users/nicholasaddison/.cargo/bin/anvil --fork-block-number 15291050 -f
```

**Comment by mattsse** [comment]
nope that's on me

will turn your example into a proper test and debug this

**Comment by naddison36** [comment]
BTW, the `from` address in the first call is now correct.

The `from` in the second `delegatecall` is incorrect, though. I noticed this yesterday but wanted to retest with https://github.com/foundry-rs/foundry/pull/2644 before raised it.

When I test against Nethermind, the `action.from` is the contract that is sending the `delegatecall` and not the sender of the transaction. That is, the `action.to` address of the first trace should be the `action.from` address of the second trace for tx [0x3255cce7312e9c4470e1a1883be13718e971f6faafb96199b8bd75e5b7c39e3a](https://etherscan.io/tx/0x3255cce7312e9c4470e1a1883be13718e971f6faafb96199b8bd75e5b7c39e3a).
You can also check out the Parity trace on Etherscan https://etherscan.io/vmtrace?txhash=0x3255cce7312e9c4470e1a1883be13718e971f6faafb96199b8bd75e5b7c39e3a&type=parity#raw

The first 2 traces from Anvil 
```json
[
        {
            "action": {
                "callType": "call",
                "from": "0x2e4777139254ff76db957e284b186a4507ff8c67",
                "gas": "0x1ea7b",
                "input": "0x43bcfab60000000000000000000000006b175474e89094c44da98b954eedeac495271d0f0000000000000000000000000000000000000000000000e0bd811c8769a824b00000000000000000000000000000000000000000000000e0ae9925047d8440b60000000000000000000000002e4777139254ff76db957e284b186a4507ff8c67",
                "to": "0xe2f2a5c287993345a840db3b0845fbc70f5935a5",
                "value": "0x0"
            },
            "blockHash": "0x57fc41478973665020c3da3e26dc0cad4e975bff441a24674e4f89c273046b54",
            "blockNumber": 15291051,
            "error": null,
            "result": {
                "gasUsed": "0x1ea7b",
                "output": "0x0000000000000000000000000000000000000000000000e0e82ca52ec6e6a4d3"
            },
            "subtraces": 1,
            "traceAddress": [
                0
            ],
            "transactionHash": "0xe3f418d4fb9ded6f13481b47ca19d3454b039c7ec96ba0778e920d39f9029a31",
            "transactionPosition": 0,
            "type": "call"
        },
        {
            "action": {
                "callType": "delegatecall",
                "from": "0x2e4777139254ff76db957e284b186a4507ff8c67",
                "gas": "0x1ce41",
                "input": "0x43bcfab60000000000000000000000006b175474e89094c44da98b954eedeac495271d0f0000000000000000000000000000000000000000000000e0bd811c8769a824b00000000000000000000000000000000000000000000000e0ae9925047d8440b60000000000000000000000002e4777139254ff76db957e284b186a4507ff8c67",
                "to": "0x15b2838cd28cc353afbe59385db3f366d8945aee",
                "value": "0x0"
            },
            "blockHash": "0x57fc41478973665020c3da3e26dc0cad4e975bff441a24674e4f89c273046b54",
            "blockNumber": 15291051,
            "error": null,
            "result": {
                "gasUsed": "0x1ce41",
                "output": "0x0000000000000000000000000000000000000000000000e0e82ca52ec6e6a4d3"
            },
            "subtraces": 2,
            "traceAddress": [
                0,
                1
            ],
            "transactionHash": "0xe3f418d4fb9ded6f13481b47ca19d3454b039c7ec96ba0778e920d39f9029a31",
            "transactionPosition": 0,
            "type": "call"
        }
```

The first two traces from Nethermind
```json
[
        {
            "action": {
                "callType": "call",
                "from": "0x2e4777139254ff76db957e284b186a4507ff8c67",
                "gas": "0x262b3",
                "input": "0x43bcfab60000000000000000000000006b175474e89094c44da98b954eedeac495271d0f0000000000000000000000000000000000000000000000e0bd811c8769a824b00000000000000000000000000000000000000000000000e0ae9925047d8440b60000000000000000000000002e4777139254ff76db957e284b186a4507ff8c67",
                "to": "0xe2f2a5c287993345a840db3b0845fbc70f5935a5",
                "value": "0x0"
            },
            "blockHash": "0xa47c8f1d8c284cb614e9c8e10d260b33eae16b1957a83141191bc335838d7e29",
            "blockNumber": 15291051,
            "result": {
                "gasUsed": "0x2131b",
                "output": "0x0000000000000000000000000000000000000000000000e0e82ca52ec6e6a4d3"
            },
            "subtraces": 1,
            "traceAddress": [],
            "transactionHash": "0x3255cce7312e9c4470e1a1883be13718e971f6faafb96199b8bd75e5b7c39e3a",
            "transactionPosition": 19,
            "type": "call"
        },
        {
            "action": {
                "callType": "delegatecall",
                "from": "0xe2f2a5c287993345a840db3b0845fbc70f5935a5",
                "gas": "0x23d88",
                "input": "0x43bcfab60000000000000000000000006b175474e89094c44da98b954eedeac495271d0f0000000000000000000000000000000000000000000000e0bd811c8769a824b00000000000000000000000000000000000000000000000e0ae9925047d8440b60000000000000000000000002e4777139254ff76db957e284b186a4507ff8c67",
                "to": "0x15b2838cd28cc353afbe59385db3f366d8945aee",
                "value": "0x0"
            },
            "blockHash": "0xa47c8f1d8c284cb614e9c8e10d260b33eae16b1957a83141191bc335838d7e29",
            "blockNumber": 15291051,
            "result": {
                "gasUsed": "0x1f6e1",
                "output": "0x0000000000000000000000000000000000000000000000e0e82ca52ec6e6a4d3"
            },
            "subtraces": 2,
            "traceAddress": [
                0
            ],
            "transactionHash": "0x3255cce7312e9c4470e1a1883be13718e971f6faafb96199b8bd75e5b7c39e3a",
            "transactionPosition": 19,
            "type": "call"
        }
```

**Comment by mattsse** [comment]
fixed smol issue with traceAddress

does match the reference tx now. will fix delegate in a separate pr

**Comment by mattsse** [comment]
this now also includes the fix for #2663

see openenthereum impl for from/to addresses

https://github.com/openethereum/openethereum/blob/6c2d392d867b058ff867c4373e40850ca3f96969/crates/ethcore/src/trace/types/trace.rs#L73-L90

File: evm/src/executor/inspector/tracer.rs [hunk 1]
```diff
@@ -84,13 +86,20 @@ where
         call: &mut CallInputs,
         _: bool,
     ) -> (Return, Gas, Bytes) {
+        let (from, to) = match call.context.scheme {
+            CallScheme::DelegateCall | CallScheme::CallCode => {
+                (call.context.address, call.context.code_address)
+            }
+            _ => (call.context.caller, call.context.address),
```
Comment (gakonst) [inline evm/src/executor/inspector/tracer.rs:93]: nice

File: anvil/core/src/eth/transaction/mod.rs [hunk 2]
```diff
@@ -1049,17 +1049,34 @@ pub struct TransactionInfo {
...
-    pub fn trace_call_graph(&self, idx: usize) -> Vec<usize> {
+    /// Returns the `traceAddress` of the node in the arena
+    ///
+    /// The `traceAddress` field of all returned traces, gives the exact location in the call trace
+    /// [index in root, index in first CALL, index in second CALL, …].
+    ///
+    /// # Panics
+    ///
+    /// if the `idx` does not belong to a node
+    pub fn trace_address(&self, idx: usize) -> Vec<usize> {
```
Comment (gakonst) [inline anvil/core/src/eth/transaction/mod.rs:1060]: 👍

**Review by gakonst — APPROVED** [review]

**Pull request merged**